### DISCUSSION TITLE
ESP8266 UDP between AP and client not working?

### DISCUSSION THREAD

**Discussion by gcanzalo** [micropython/micropython#10148] [Using MicroPython] [answered]
Hi:

MicroPython v1.19.1 on 2022-06-18; ESP module with ESP8266 running on a pair of D1 minis. One is set as an AP:

```py
import socket
import select
import network
from time import sleep, time

import select

ap = network.WLAN(network.AP_IF)
ap.active(True)
ap.ifconfig(('192.168.4.1', '255.255.255.0', '192.168.4.1', '192.168.4.1'))
ap.config(essid = 'esp8266wifi', password = '123456789')

while ap.active() == False:
  sleep(1)
  print('starting...')

print('Connection successful')
print(ap.ifconfig())

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('192.168.4.1', 6060))

while True:
    sleep(0.01)
    readable, writeable, exceptional = select.select([sock], [sock], [sock], 1)

    for s in readable:
      print(s.recvfrom(8))
```

The other as a client:

```py
import socket
import network
from time import sleep, time
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.ifconfig(('192.168.4.10', '255.255.255.0', '192.168.4.1', '192.168.4.1'))
wlan.connect('esp8266wifi', '123456789')
while not wlan.isconnected():
    print('connecting to network...')
    sleep(1)

print('network config:', wlan.ifconfig())

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
# sock.bind(('192.168.4.10', 12345))

while True:
    sleep(1)
    print(sock.sendto(b'1', ('192.168.4.1', 6060)))
    print(wlan.ifconfig())
```

Which doesn't work. The client sends, but the AP doesn't receive the packet. If I uncomment the bind statement in the client script, an EINVAL error is thrown on sendto. The AP receives packets fine from my PC if I run the following:

```py
import socket
from time import sleep, time

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('', 6060))

while True:
    sleep(1)
    print(sock.sendto(b'from PC', ('192.168.4.1', 6060)))
```

What am I doing wrong?

Thank you.

**Comment by davefes** [comment, marked as answer]
I have found that setting static IPs on the ESP32 needed a short delay, say 0.5 seconds before it.  Throwing some try/except blocks around might give you a clue.

**Reply by davefes** [reply]
That:
```py
# sock.bind(('192.168.4.10', 12345))
```
are you sure that is what you want?

**Reply by gcanzalo** [reply]
The client bind attempt was just me trying to understand what binding actually does; it's not what I want.

The timing clue put me on the right path. It appears actually to be an order of operations issue. If ifconfig() and config() are called prior to active(), everything works even without a pause. This is the issue with strictly following the tutorials, which suggest order of operation isn't important...

Thanks!

**Reply by davefes** [reply]
Interesting ... which tutorial(s)?  I thought it had to be active first then configure.

My connect():

```py
    if sta_if.isconnected():
        sta_if.disconnect()
        print ('started in the connected state')
        print ('but now it is disconnected')
    else:
        print ('started in the disconnected state')

    utime.sleep(1)

    if not sta_if.isconnected():
        print('connecting to hotspot...')
        sta_if.active(True)
        utime.sleep(.5) #  or you get a crash
        sta_if.ifconfig((config.WiFi_device, '255.255.255.0', config.gateway, '8.8.8.8'))

        try:
            sta_if.connect(config.hotspot, config.password)
        except OSError as error:
            try:
                with open('errors.txt', 'a') as outfile:
                    outfile.write(error + '\n')
            except OSError:
                pass


        while (count < 5):
            count += 1

            if (sta_if.isconnected()):
                count = 0
                print (' network config:', sta_if.ifconfig())
                break

            print ('.', end = '')
            utime.sleep(1)
```
The first part was only relevant for ESP32 and I think it has been fixed.  My 0.5 second "fix" before setting a static IP may have been fixed on ESP32.

**Reply by gcanzalo** [reply]
I perused several tutorials, all had active() before config(), so I assumed (emphasis on the ass) active() was to be called prior configuration. Not so. Your script should work fine without sleep() simply by moving active() after all configs.

I'm quite pleased with the result, getting ~160 Hz transmission rates of 8 byte packets between these two cheap, little devices! On my way to having a wireless TIG foot pedal!

**Reply by davefes** [reply]
Do you mind if I elevated this to try and get a common understanding on whether or not you config or make active first?  I have spent countless hours trying to get bullet-proof WiFi connections even with hotspots a few metres away.

I am working on an ESP32-CAM and he inits first and then changes camera settings.
A similar camera effort here:
https://randomnerdtutorials.com/esp32-cam-send-photos-email/
inits first and then changes camera settings

As you point out the docs:
https://docs.micropython.org/en/latest/esp32/quickref.html
suggest that making the interface active is the first thing you do.

**Reply by gcanzalo** [reply]
Elevate away...

**Reply by gcanzalo** [reply]
This is quite puzzling. I just cleared the flash with esptool, flashed the latest stable micropython, uploaded my script with the ap.active() after the ap.ifconfig(), and the script raised:

```
OSError: wifi_set_ip_info() failed
```
Moved ap.active() _before_ ap.ifconfig() and it ran without exception. Then moved the ap.active() back where it was, _after_ ap.ifconfig(), and it again runs fine. WTF?

**Reply by davefes** [reply]
Unfortunately, I didn't have a try/except around:
`sta_if.ifconfig((config.WiFi_device, '255.255.255.0', config.gateway, '8.8.8.8'))
`
or I might have seen that error.  I didn't get the answer I wanted in:
https://github.com/orgs/micropython/discussions/10152
